runtime/eventring.c

```diff
@@ -0,0 +1,543 @@
+The producer code is contained here . By default a <pid>.eventring file is
+created in the current directory (overridable by setting
+OCAML_EVENTRING_DIR). This file contains a ring buffer for each possible domain
+(Max_domains). It is laid out in a structure that enables sparsity. On-disk
+(or in-memory) footprint is proportional to the max number of concurrent domains
+the process has ever run.
+
+On disk structure:
+
+----------------------------------------------------------------
+| File header (version, offsets, etc..)                        |
+----------------------------------------------------------------
+| Ring 0..Max_domains metadata                                 |
+| (head and tail indexes, one per cache line)                  |
+----------------------------------------------------------------
+| Ring 0..Max_domains data                                     |
+| (actual ring data, default 2^16 words = 512k bytes)          |
+----------------------------------------------------------------
+
+*/
+
+typedef enum { EV_RUNTIME, EV_USER } ev_category;
+
+/* These store state for the current ring buffers open for writing */
+static struct eventring_metadata_header *current_metadata = NULL;
+static int current_ring_total_size;
+static char_os *eventring_path;
+static char_os *current_ring_loc = NULL;
+
+#ifdef _WIN32
+static HANDLE ring_file_handle;
+static HANDLE ring_handle;
+#endif
+
+/* This comes from OCAMLRUNPARAMs and is initialised in caml_eventring_init */
+static int ring_size_words;
+
+static atomic_uintnat eventring_enabled = 0;
+static atomic_uintnat eventring_paused = 0;
+
+static void write_to_ring(ev_category category, ev_message_type type,
+                          int event_id, int event_length, uint64_t *content,
+                          int word_offset);
+
+void caml_eventring_init() {
+  eventring_path = caml_secure_getenv(T("OCAML_EVENTRING_DIR"));
+
+  ring_size_words = 1 << caml_params->eventring_size;
+
+  if (caml_secure_getenv(T("OCAML_EVENTRING_START"))) {
+    caml_eventring_start();
+  }
+}
+
+/* Remove the ringbuffers. This is only to be called in a stop-the-world */
+static void teardown_eventring(caml_domain_state *domain_state, void *data,
+                               int num_participating,
+                               caml_domain_state **participating_domains) {
+  caml_global_barrier();
+  if (participating_domains[0] == domain_state) {
+#ifdef _WIN32
+    UnmapViewOfFile(current_metadata);
+    CloseHandle(ring_file_handle);
+    CloseHandle(ring_handle);
+#else
+    munmap(current_metadata, current_ring_total_size);
+
+    int remove_file = *(int*)data;
+
+    if( remove_file ) {
+      unlink(current_ring_loc);
+    }
+#endif
+
+    caml_stat_free(current_ring_loc);
+    current_metadata = NULL;
+
+    atomic_store_rel(&eventring_enabled, 0);
+  }
+  caml_global_barrier();
+}
+
+void caml_eventring_post_fork() {
+  /* We are after a call to fork (which can only happen when there is a single
```

Comment:
if I understand correctly, we are running in the child (not in the parent); this could be specified.